pkg/operator/configobservation/cloudprovider/observe_cloudprovider.go

```diff
@@ -0,0 +1,130 @@
+	configNamespace           = "openshift-config"
+)
+
+func ObserveCloudProviderNames(
+	genericListers configobserver.Listers,
+	recorder events.Recorder,
+	existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	var errs []error
+	cloudProvidersPath := []string{"extendedArguments", "cloud-provider"}
+	cloudProviderConfPath := []string{"extendedArguments", "cloud-config"}
+	previouslyObservedConfig := map[string]interface{}{}
+
+	existinCloudConfig, _, err := unstructured.NestedStringSlice(existingConfig, cloudProviderConfPath...)
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	if currentCloudProvider, _, _ := unstructured.NestedStringSlice(existingConfig, cloudProvidersPath...); len(currentCloudProvider) > 0 {
+		if err := unstructured.SetNestedStringSlice(previouslyObservedConfig, currentCloudProvider, cloudProvidersPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(existinCloudConfig) > 0 {
+		if err := unstructured.SetNestedStringSlice(previouslyObservedConfig, existinCloudConfig, cloudProviderConfPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	observedConfig := map[string]interface{}{}
+
+	infrastructure, err := listers.InfrastructureLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		recorder.Warningf("ObserverCloudProviderNames", "Required infrastructures.%s/cluster not found", configv1.GroupName)
+		return observedConfig, errs
+	}
+	if err != nil {
+		return previouslyObservedConfig, errs
+	}
+
+	cloudProvider := getPlatformName(infrastructure.Status.Platform, recorder)
+	if len(cloudProvider) > 0 {
+		if err := unstructured.SetNestedStringSlice(observedConfig, []string{cloudProvider}, cloudProvidersPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	sourceCloudConfigMap := infrastructure.Spec.CloudConfig.Name
+	sourceCloudConfigNamespace := configNamespace
+	sourceLocation := resourcesynccontroller.ResourceLocation{
+		Namespace: sourceCloudConfigNamespace,
+		Name:      sourceCloudConfigMap,
+	}
+
+	if len(sourceCloudConfigMap) == 0 {
+		sourceLocation = resourcesynccontroller.ResourceLocation{}
+	}
+
+	err = listers.ResourceSyncer().SyncConfigMap(
+		resourcesynccontroller.ResourceLocation{
+			Namespace: targetNamespaceName,
+			Name:      "cloud-config",
+		},
+		sourceLocation)
+
+	if err != nil {
+		errs = append(errs, err)
+		return observedConfig, errs
+	}
+
+	if len(sourceCloudConfigMap) == 0 {
+		return observedConfig, errs
+	}
+
+	// usually key will be simply config but we should refer it just in case
+	staticCloudConfFile := fmt.Sprintf(cloudProviderConfFilePath, infrastructure.Spec.CloudConfig.Key)
+
+	if err := unstructured.SetNestedStringSlice(observedConfig, []string{staticCloudConfFile}, cloudProviderConfPath...); err != nil {
+		recorder.Warningf("ObserverCloudProviderNames", "Failed setting cloud-config : %v", err)
```

Comment:
s/ObserverCloudProviderNames/ObserverCloudProviderNamesFailed/

(usually reason is always Failed/Succeeded/Changed/etc.)